services/gundeck/test/unit/MockGundeck.hs

```diff
@@ -0,0 +1,712 @@
+        insertAllClients (same, (Recipient uid route [], pay)) = [(same, (rcp', pay))]
+          where rcp' = Recipient uid route (clientIdsOfUser env uid)
+
+        -- otherwise, no special hidden meaning.
+        insertAllClients same@(_, (Recipient _ _ (_:_), _)) = [same]
+
+    expectNative :: MockEnv -> NotifQueue
+    expectNative env
+      = foldl' (uncurry . deliver) mempty
+      . filter reachable
+      . reformat
+      . mconcat . fmap removeSome
+      . mconcat . fmap insertAllClients
+      $ rcps
+      where
+        reachable :: ((UserId, ClientId), payload) -> Bool
+        reachable (ids, _) = nativeReachable env ids && not (wsReachable env ids)
+
+        removeSome :: ((UserId, Maybe ClientId, Bool), (Recipient, Payload)) -> [(Recipient, Payload)]
+        removeSome ((_, _, _), (Recipient _ RouteDirect _, _)) =
+          -- never native-push recipients marked 'RouteDirect'.
+          []
+        removeSome ((snduid, _, False), same@(Recipient rcpuid _ _, _)) =
+          -- if pushNativeIncludeOrigin is False, none of the originator's devices will receive
+          -- native pushes.
+          [same | snduid /= rcpuid]
+        removeSome ((_, Just sndcid, True), (Recipient rcpuid route cids, pay)) =
+          -- if originating client is known and pushNativeIncludeOrigin is True, filter out just
+          -- that client, and native-push to all other devices of the originator.
+          [(Recipient rcpuid route $ filter (/= sndcid) cids, pay)]
+        removeSome ((_, Nothing, True), same) =
+          -- if NO originating client is known and pushNativeIncludeOrigin is True, push to all
+          -- originating devices.
+          [same]
+
+        insertAllClients :: (any, (Recipient, Payload))
+                         -> [(any, (Recipient, Payload))]
+        -- if the recipient client list is empty, fill in all devices of that user
+        insertAllClients (same, (Recipient uid route [], pay)) = [(same, (rcp', pay))]
+          where rcp' = Recipient uid route (clientIdsOfUser env uid)
+        -- otherwise, no special hidden meaning.
+        insertAllClients same@(_, (Recipient _ _ (_:_), _)) = [same]
+
+    expectCass :: NotifQueue
+    expectCass = foldl' (uncurry . deliver) mempty . mconcat $ go <$> pushes
+      where
+        go :: Push -> [((UserId, ClientId), Payload)]
+        go psh =
+          [ ((uid, cid), psh ^. pushPayload)
+          | not $ psh ^. pushTransient
+          , Recipient uid _ cids <- toList . fromRange $ psh ^. pushRecipients
+          , cid <- cids
+          ]
+
+    reformat :: [(Recipient, Payload)] -> [((UserId, ClientId), Payload)]
+    reformat = mconcat . fmap go
+      where
+        go (Recipient uid _ cids, pay) = (\cid -> ((uid, cid), pay)) <$> cids
+
+    rcps :: [((UserId, Maybe ClientId, Bool), (Recipient, Payload))]
+    rcps = mconcat $ go <$> filter (not . (^. pushTransient)) pushes
```

Comment:
Why do we filter out transient pushes? Shouldn't they still be pushed via websockets?

Reply:
i think that's a bug that wasn't caught because genPush didn't touch it.  you're thorough with this one!  :-)